### DISCUSSION TITLE
addComponent in plugin function doesn't seem to work anymore?

### DISCUSSION THREAD

**Discussion by dludemann** [tailwindlabs/tailwindcss#10383] [Help]
Not sure if I was exploiting a bug or something but here is my situation. When TW3 released I had created a plugin that was used to overwrite some styles from react component library. For the sake of ease I'm keeping the terms generic.

I have a component library (BobUI) that has an input component like this
```html
<input class="bob-input"/>
```

To use it in my react project I import and use like so:
```javascript
import BobInput from 'BobUI'
<BobInput />
```

I wrote a plugin to add / overwrite some of those styles:
```javascript
    plugins: [
        plugin(function ({ addComponents }) {
            addComponents({
                '.bob-input': {
                    '@apply h-20 w-96': {},
                },
            });
        }),
    ],
```

Now this **use** to work for me. .bob-input was injected into the styles and the styles were overwritten. I recently started a new project with the same code and now it doesn't work. 

I notice now that the documentation says 
```
Like with other component classes in Tailwind, component classes added by a plugin will only be included in the generated CSS if they are actually being used in the project.
```

which I am interpreting as ".bob-input has to be found in one of the files in your `content: []` array".

Which I did the following:
content: ["node_modules/bob-ui/src/**/*.{js,ts,jsx,tsx}"]

This line makes it work. But why did it work before without this? The problem with this is now any tailwind class this library uses or if they have a class with the same name as a tailwind class it is now getting bundled in the css. 

A way around this is to re-declare bob-input in the actual project.
```
<BobInput className="bob-input" />
```

this is redundant though. Is there not a way to use Tailwind plugins to add classes that will be generated regardless if it's being used in the project?  Or do I have to safelist everything used in the plugin?
```

**Comment by wongjn** [comment]
> Is there not a way to use Tailwind plugins to add classes that will be generated regardless if it's being used in the project? Or do I have to safelist everything used in the plugin?

There is a way to use Tailwind plugins to add classes that will be generated regardless if it's being used in the project without safelist – use an attribute selector, like `[class~="bob-input"]`.

**Reply by dludemann** [reply]
This is exactly what I was looking for, thank you!

**Comment by adamwathan** [comment]
Hey! Not sure what the root of the issue is for you here but this behavior hasn't changed at all since the initial release of v3 — anything added using `addComponents` has always only ever been inserted if the classes were actually used. You can install `tailwindcss@3.0.0` and try it out and you'll see the behavior is the same.

What I would recommend is that consumers of your library included the path to your library's template files in their own config:

```js
module.exports = {
  content: ['./node_modules/bobui/bobui/src/**/*.html']
  // ...
}
```

...or whatever the path is 👍
